### PR TITLE
[facerecognition] Face Recognition Binding initial contribution

### PR DESCRIPTION
**Pull request by philippmeisberger**
This new binding receives images from IP cameras and provides user authentication through face recognition. The recognition is done by OpenCV.

### CODE REVIEW

**Comment by martinvw** [comment]
Thanks for your PR! Sounds awesome.

Please include the binding in the parent `pom.xml` in that case it will already be tested and analysed by travis. And it will become available in [jfrog builds list](https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/binding/).

Please also check:

http://docs.openhab.org/developers/development/bindings.html
http://docs.openhab.org/developers/development/guidelines.html

And it we would be great if you could also review the report of [the Static code analysis](https://docs.openhab.org/developers/development/bindings.html#static-code-analysis), this also saves me some time while reviewing.

**Comment by philippmeisberger** [comment]
I already included the parent in the `pom.xml` and also checked the guidelines ;) 
The OpenCV jars are required for building. I keep them locally. Should I commit them too?

**Comment by philippmeisberger** [comment]
How do I execute the static code analysis? Is it maybe executed automatically if I run `mvn clean package`?

**Comment by martinvw** [comment]
You included the parent in the Pom.xml but not your new binding in the parent Pom.xml (in the bindings folder)

Btw @kaikreuzer should this be a binding or something else io...?

**Comment by philippmeisberger** [comment]
Oh yes! you are right!

**Comment by martinvw** [comment]
And now Travis will also report your build errors 🙂

So yes, you have to include the jar if it’s required to build it, you should put it in the lib folder.

Clean install will give you the report for sure, I don’t know in which exact maven phase it’s included.

**Comment by philippmeisberger** [comment]
Sry, but I forgot the "Signed-off-by" in the last commit...

Poorly the jars are architecture dependent... I already tried to just add a maven dependency but the packages are downloaded into `$HOME/.m2/`. They should be downloaded into `lib` directory. I am newbie with using maven... Maybe you have a suggestion? I use the jars from https://github.com/bytedeco/javacv

**Comment by kaikreuzer** [comment]
FIrst of all: Looks like a great feature, @philippmeisberger!

> Btw @kaikreuzer should this be a binding or something else io...?

I would think so. It does not really qualify as a kind of "device integration"; actually, it should not even access cameras itself, this would rather be done by bindings like https://github.com/openhab/openhab2-addons/pull/2282.

If I understand it correctly, it is a "piece of configurable logic", which takes an Image item as an input and provides a username & score as an output (roughly spoken).
This could imho be a perfect fit for a rule action (ideally for the new rule engine), wdyt?

**Comment by philippmeisberger** [comment]
So should I better use namespace `org.openhab.io.facerecognition`?
It's a great idea to just use the #2282... poorly I spent much time in this binding to get it working. It would take some time to modify it...

You are right: The binding has a configurable logic which is based on machine learning. Users must be trained to be authenticated. The current user face model is compared against the trained (stored) model. This results in a _score_. There is the _threshold_ configuration parameter which determines if the user is accepted or not. How do you mean this could be ideally for the new rule engine?

**Comment by kaikreuzer** [comment]
>  How do you mean this could be ideally for the new rule engine?

Not sure how much you already know about the new rule engine, but you can find some info here: https://www.eclipse.org/smarthome/documentation/features/rules.html
What I tried to say is that this feature could be implemented as an "action module type" - which is effectively some logic for which you define inputs and outputs and configuration parameters.

> So should I better use namespace org.openhab.io.facerecognition?

We do not yet have this "rule module add-on" category formalised for openHAB - there aren't yet any such add-ons available. The first prototype is done in https://github.com/openhab/openhab2-addons/pull/2926, so `org.openhab.action` might be the namespace to use in the end.

> It's a great idea to just use the #2282... poorly I spent much time in this binding to get it working. It would take some time to modify it...

I would appreciate if you would invest this time, so that we end up with a clean and future-proof solution for it. Maybe you can provide valuable input on #2282 with all the experience that you have gathered in this regard?

**Comment by clinique** [comment]
Don't you think this binding could be made more generic using profiles and taking Image type item as input, removing IP camera access code ?

**Comment by philippmeisberger** [comment]
Yes, this should be possible. Images can be published by another binding (e.g. IP camera), right? So this binding should receive those. All known users must be added manually to configuration anyway. I could take a look at this and rewrite the binding this way but only if the OpenHab team likes this suggestion ;)

**Comment by martinvw** [comment]
Luckily @kaikreuzer already answered this, a year ago:

> I would think so. It does not really qualify as a kind of "device integration"; actually, it should not even access cameras itself, this would rather be done by bindings like #2282.
>
> If I understand it correctly, it is a "piece of configurable logic", which takes an Image item as an input and provides a username & score as an output (roughly spoken).
> This could imho be a perfect fit for a rule action (ideally for the new rule engine), wdyt?

**Comment by davidgraeff** [comment]
This addon classifies as binding, because it is accessing external resources. (Otherwise we can start to remove all the weather bindings ;) )

Additionally an action handler like lgwebos or the mqtt binding (and other bindings by now) offer it would be nice to have.

But I do not see a reason to block this contribution and actually would like to have that functionality available.

The mentioned camera contribution as a base, similar to the MQTT concept, is a fine idea. You would configure your IP camera in that binding and then add a face-recognition thing on top of that camara Bridge. But it concerns exactly one file with about 50 lines (see https://github.com/openhab/openhab2-addons/pull/3329/files#diff-c8cbb5e381c37bff932f079e331b5145R155) and a refactoring is very easily doable after the other PR made progress. (It doesn't look like it, although just 500 lines!)

I'll perform a review now.

File: addons/binding/org.openhab.binding.facerecognition/.classpath [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.facerecognition/.classpath]: Could you please migrate to the new buildsystem? Have a look at the sticky issue for more information.

File: addons/binding/org.openhab.binding.facerecognition/ESH-INF/thing/thing-types.xml [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+			<channel-group typeId="live-group" id="live"></channel-group>
+			<channel-group typeId="accessGranted-group" id="accessGranted"></channel-group>
+			<channel-group typeId="accessDenied-group" id="accessDenied"></channel-group>
+		</channel-groups>
+		<config-description>
+			<parameter name="camera" type="text">
+				<required>true</required>
+				<context>network_address</context>
+				<description>URL for receiving camera snapshots</description>
+				<default>http://IP/image.jpg</default>
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.facerecognition/ESH-INF/thing/thing-types.xml]: Better not do that. The default value will satisfy the "required" attribute, but the default is not a valid address.

File: addons/binding/org.openhab.binding.facerecognition/ESH-INF/thing/thing-types.xml [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+				<required>true</required>
+				<default>2000</default>
+				<description>Interval in milliseconds between the captures</description>
+			</parameter>
+			<parameter name="username" type="text">
+				<description>User name for accessing camera</description>
+				<default></default>
+			</parameter>
+			<parameter name="password" type="text">
+				<description>Password for accessing camera</description>
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.facerecognition/ESH-INF/thing/thing-types.xml]: Please add `<context>password</context>`

File: addons/binding/org.openhab.binding.facerecognition/README.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+The file `haarcascade_frontalface_default.xml` from the package `opencv-data` is required to detect faces. On Debian/Ubuntu based systems just install that package:
+
+    ~# apt-get install opencv-data
+
+On other systems copy that file to `/usr/share/opencv/haarcascades/` directory.
+
+## Discovery
+
+Discovery is currently not supported.
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.facerecognition/README.md]: Please remove this section

File: addons/binding/org.openhab.binding.facerecognition/README.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+## Binding Configuration
+
+No binding configuration required.
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.facerecognition/README.md]: Please remove this section

File: addons/binding/org.openhab.binding.facerecognition/about.html [hunk 6]
```diff
@@ -0,0 +1,32 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.facerecognition/about.html]: Please update to the NOTICE file of eclipse license 2 (see other bindings)

File: addons/binding/org.openhab.binding.facerecognition/lib/README.md [hunk 7]
```diff
@@ -0,0 +1,4 @@
+Download required libraries
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.facerecognition/lib/README.md]: When you update to the new buildsystem, please remove the "lib" directory. You can finally declare dependencies in your pom.xml file. Have a look at jcenter or maven central and find the required dependencies.

File: addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+/**
+ * The {@link FaceRecognitionHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Philipp Meisberger - Initial contribution
+ */
+public class FaceRecognitionHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(FaceRecognitionHandler.class);
+    private final ScheduledExecutorService service = Executors.newSingleThreadScheduledExecutor();
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java]: please use "scheduler" which is made available to you by the base thing handler.

File: addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+    private final ChannelUID imageChannelUID;
+    private final ChannelUID accessGrantedImageChannelUID;
+    private final ChannelUID accessGrantedUsernameChannelUID;
+    private final ChannelUID accessGrantedScoreChannelUID;
+    private final ChannelUID accessGrantedTimeChannelUID;
+    private final ChannelUID accessDeniedImageChannelUID;
+    private final ChannelUID accessDeniedUsernameChannelUID;
+    private final ChannelUID accessDeniedScoreChannelUID;
+    private final ChannelUID accessDeniedTimeChannelUID;
+    private final static String MODELS_FILE = "/var/lib/openhab2/etc/org.openhab.binding.facerecognition.models.xml";
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java]: Please make this a configuration option with a default value. But this path is not the same across operating systems or installations.

File: addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+    private final static String FACE_CLASSIFIER_FILE = "/usr/share/opencv/haarcascades/haarcascade_frontalface_default.xml";
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java]: Please make this a configuration option

File: addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link FaceRecognitionHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Philipp Meisberger - Initial contribution
+ */
+public class FaceRecognitionHandler extends BaseThingHandler {
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java]: Please annotate your classes with `@NonNullByDefault` which prevents a lot of NPEs.

File: addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+    private final static String FACE_CLASSIFIER_FILE = "/usr/share/opencv/haarcascades/haarcascade_frontalface_default.xml";
+    private ScheduledFuture<?> job;
+    private CascadeClassifier faceCascade;
+    private Mat accessDeniedFace;
+    private BigDecimal interval;
+    private BigDecimal threshold;
+
+    public FaceRecognitionHandler(Thing thing) {
+        super(thing);
+        httpClient = new HttpClient();
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java]: Use the http client factory service from `org.eclipse.smarthome.core.net`.
You need to acquire that service in your factory and hand it to your thing handler as a parameter

File: addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+                faceRecognizer.setLabelInfo(label, command.toString());
+                faceRecognizer.save(MODELS_FILE);
+                logger.debug("Trained face of user \"{}\"", command.toString());
+            }
+        }
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initializing {}::{}", getThing().getLabel(), getThing().getUID());
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java]: You don't need to log infront of an `updateStatus`. The framework will tell you about those events already

File: addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java [hunk 14]
```diff
@@ -0,0 +1,310 @@
...
+
+        } catch (Exception e) {
+            logger.error("{}: {}", e.getClass().getName(), e.getMessage());
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+        }
+    }
+
+    @Override
+    public void dispose() {
+        logger.debug("Disposing {}::{}", getThing().getLabel(), getThing().getUID());
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java]: No need to log that. The framework is logging this as well.

File: addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java [hunk 15]
```diff
@@ -0,0 +1,310 @@
...
+        if (!job.isDone()) {
+            logger.debug("Cancelling snapshot job");
+            job.cancel(true);
+        }
+
+        try {
+            httpClient.stop();
+
+        } catch (Exception e) {
+            logger.error("{}", e.getMessage());
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java]: Don't use the error level that is reserved to the framework. The last argument has to be the exception itself to get the stacktrace. The line should read: `logger.warn(e);`

File: addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java [hunk 16]
```diff
@@ -0,0 +1,310 @@
...
+
+        if (!job.isDone()) {
+            logger.debug("Cancelling snapshot job");
+            job.cancel(true);
+        }
+
+        try {
+            httpClient.stop();
+
+        } catch (Exception e) {
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java]: Don't catch generic exceptions

File: addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java [hunk 17]
```diff
@@ -0,0 +1,310 @@
...
+                // User recognized?
+                if (facesTrained && (score.get(0) <= threshold.intValue())) {
+
+                    if (isLinked(CHANNEL_ACCESS_GRANTED_IMAGE)) {
+                        ByteBuffer outputBuffer = ByteBuffer.allocate(face.arraySize());
+                        imencode(".jpg", face, outputBuffer);
+                        updateState(accessGrantedImageChannelUID, new RawType(outputBuffer.array(), "image/jpeg"));
+                    }
+
+                    if (isLinked(CHANNEL_ACCESS_GRANTED_USERNAME)) {
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java]: You really don't need to check for a channel if it is linked. Just update it. Only if it costs a lot of resources to compute a channel (like with the `CHANNEL_ACCESS_GRANTED_IMAGE` one) you might check for a linkage.

File: addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java [hunk 18]
```diff
@@ -0,0 +1,310 @@
...
+                    if (isLinked(CHANNEL_ACCESS_DENIED_TIME)) {
+                        updateState(accessDeniedTimeChannelUID, new DateTimeType());
+                    }
+                }
+            }
+
+            updateStatus(ThingStatus.ONLINE);
+
+        } catch (Exception e) {
+            logger.error("{}: {}", e.getClass().getName(), e.getMessage());
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java]: logger.error. That's also not necessary infront of `updateStatus`. And you are catching a generic exception here.

**Comment by kaikreuzer** [comment]
> This addon classifies as binding

@davidgraeff I think you have missed the above discussion.
This contribution mixes two things:
1. Retrieving pictures from IP cameras (this is what should be merged with #2282)
2. Mapping images to persons - this is exactly what profiles are about as @clinique mentioned.

Part (2) is the interesting one here in this contribution. And having it implemented as a profile would allow ANY binding that has Image channels to make use of this feature.

**Comment by davidgraeff** [comment]
I got it. Profiles are mostly implemented in transformation bundles as far as I have seen.

So the idea is that the configuration of the input source would haven via Things,
but that this binding should(?) also offer profiles for the configured input sources.

That can be added as a follow up PR, I guess? (Next to adding the action handler)

**Comment by philippmeisberger** [comment]
> Part (2) is the interesting one here in this contribution. And having it implemented as a profile would allow ANY binding that has Image channels to make use of this feature.

@kaikreuzer IMHO this would be great. How should this work in the end? Users must be trained anyway to be recognized. There must be some UI anywhere containing a button and edit field for this. I cannot imagine how this could be achieved without a binding. Some hints could help me to understand how you imagine this.

**Comment by davidgraeff** [comment]
https://github.com/tzolov/mtcnn-java

No model training required, the Tensorflow models are prebundled. opencv is nice, but in combination with tensorflow it really rocks.

**Comment by philippmeisberger** [comment]
> https://github.com/tzolov/mtcnn-java (for general face recognition. not sure if the detected face landmarks are enough to identify a person)
> 
> No model training required, the Tensorflow models are prebundled. opencv is nice, but in combination with tensorflow it really rocks.

@davidgraeff The training is required so that users can be recognized. The project you have found aims to just detect faces. My binding is shipped with pre-trained face detection parameters (thanks to OpenCV). The binding first detects the face and than tries to recognize it using the trained user face model. This model has to be trained by users to let their faces be recognized by my binding.

**Comment by clinique** [comment]
@philippmeisberger : hello. Do you plan to progress on this PR ?

**Comment by philippmeisberger** [comment]
@clinique Hi. I am still interested in this PR but kind of confused right now of what I should do now.

**Comment by clinique** [comment]
Well, I think first you need to rebase your PR as openhab-core as evolved quite a lot since it was proposed. But maybe even better restart from scratch with a profile type orientation.
You can find some examples of such PRs here (in openhab-core) : 
[here](https://github.com/openhab/openhab-core/pull/744/)
[or here](https://github.com/openhab/openhab-core/pull/757)
[or there](https://github.com/openhab/openhab-core/pull/779)

It will simplify your binding as you'll have no more to take care of the ipcamera part.
The tricky thing I see is how to handle unknown faces handlings because profiles can not receive informations. But I'm sure we'll find a convenient way to do it.

**Comment by clinique** [comment]
@philippmeisberger : I made an initial WIP PR in openhab-core based on your work. If you want to jump in, you're more than welcome.

**Comment by philippmeisberger** [comment]
@clinique That's great! Thanks for your invitation!

**Comment by clinique** [comment]
You’re more than welcome because you did all groundwork and I just transposed it, now hard work starts

**Comment by J-N-K** [comment]
Is this still active if work based on this one is now going on in core?

**Comment by Hilbrand** [comment]
Closing this pr due to long time no activity. It still would be interesting to have such features available, but the work in core also seem to have stalled.

**Pull request closed without merging**